src/grpc/infer_handler.h

```diff
@@ -859,7 +990,9 @@ class InferHandlerState {
 
   std::shared_ptr<Context> context_;
   Steps step_;
-  std::mutex step_mtx_;
+  std::recursive_mutex step_mtx_;
```

Comment:
Does this have to be recursive? Is it just to be able to cancel earlier while lock is already held?

Reply:
We can not release the state object if the InferResponseComplete is not yet called. The cancelled requests still invokes the Callback with cancelled response. And as state objects are reuse it is possible that InferResponseComplete of cancelled request receives response of a next request.
An alternative is to pass the current state object whose mutex is already held and skip holding the lock for that particular state when iterating through inflight states. But that seems little ugly and recursive locks fit the use-case.